soc/riscv/riscv-ite/common/chip_chipregs.h

```diff
@@ -1625,55 +1625,57 @@
+#define IT8XXX2_ADC_INTDVEN			BIT(5)
+#define IT8XXX2_ADC_KDCTL        ECREG(IT8XXX2_ADC_BASE + 0x05)
+#define IT8XXX2_ADC_AHCE				BIT(7)
+#define IT8XXX2_ADC_VCH1CTL      ECREG(IT8XXX2_ADC_BASE + 0x06)
+#define IT8XXX2_ADC_VCH1DATL     ECREG(IT8XXX2_ADC_BASE + 0x07)
+#define IT8XXX2_ADC_VCH1DATM     ECREG(IT8XXX2_ADC_BASE + 0x08)
+#define IT8XXX2_ADC_VCH2CTL      ECREG(IT8XXX2_ADC_BASE + 0x09)
+#define IT8XXX2_ADC_VCH2DATL     ECREG(IT8XXX2_ADC_BASE + 0x0A)
+#define IT8XXX2_ADC_VCH2DATM     ECREG(IT8XXX2_ADC_BASE + 0x0B)
+#define IT8XXX2_ADC_VCH3CTL      ECREG(IT8XXX2_ADC_BASE + 0x0C)
+#define IT8XXX2_ADC_VCH3DATL     ECREG(IT8XXX2_ADC_BASE + 0x0D)
+#define IT8XXX2_ADC_VCH3DATM     ECREG(IT8XXX2_ADC_BASE + 0x0E)
+#define IT8XXX2_ADC_VHSCDBL      ECREG(IT8XXX2_ADC_BASE + 0x14)
+#define IT8XXX2_ADC_VHSCDBM      ECREG(IT8XXX2_ADC_BASE + 0x15)
+#define IT8XXX2_ADC_VCH0DATL     ECREG(IT8XXX2_ADC_BASE + 0x18)
+#define IT8XXX2_ADC_VCH0DATM     ECREG(IT8XXX2_ADC_BASE + 0x19)
+#define IT8XXX2_ADC_VHSGCDBL     ECREG(IT8XXX2_ADC_BASE + 0x1C)
+#define IT8XXX2_ADC_VHSGCDBM     ECREG(IT8XXX2_ADC_BASE + 0x1D)
+#define IT8XXX2_ADC_ADCSAR       ECREG(IT8XXX2_ADC_BASE + 0x32)
+#define IT8XXX2_ADC_VCMPSCP      ECREG(IT8XXX2_ADC_BASE + 0x37)
+#define IT8XXX2_ADC_VCH4CTL      ECREG(IT8XXX2_ADC_BASE + 0x38)
 /* Voltage channel enable (ch4~ch7) */
-#define IT83XX_ADC_VCHEN			BIT(4)
-#define IT83XX_ADC_VCH4DATM     ECREG(IT83XX_ADC_BASE + 0x39)
-#define IT83XX_ADC_VCH4DATL     ECREG(IT83XX_ADC_BASE + 0x3A)
-#define IT83XX_ADC_VCH5CTL      ECREG(IT83XX_ADC_BASE + 0x3B)
-#define IT83XX_ADC_VCH5DATM     ECREG(IT83XX_ADC_BASE + 0x3C)
-#define IT83XX_ADC_VCH5DATL     ECREG(IT83XX_ADC_BASE + 0x3D)
-#define IT83XX_ADC_VCH6CTL      ECREG(IT83XX_ADC_BASE + 0x3E)
-#define IT83XX_ADC_VCH6DATM     ECREG(IT83XX_ADC_BASE + 0x3F)
-#define IT83XX_ADC_VCH6DATL     ECREG(IT83XX_ADC_BASE + 0x40)
-#define IT83XX_ADC_VCH7CTL      ECREG(IT83XX_ADC_BASE + 0x41)
-#define IT83XX_ADC_VCH7DATM     ECREG(IT83XX_ADC_BASE + 0x42)
-#define IT83XX_ADC_VCH7DATL     ECREG(IT83XX_ADC_BASE + 0x43)
-#define IT83XX_ADC_ADCDVSTS     ECREG(IT83XX_ADC_BASE + 0x44)
+#define IT8XXX2_ADC_VCHEN			BIT(4)
+#define IT8XXX2_ADC_VCH4DATM     ECREG(IT8XXX2_ADC_BASE + 0x39)
+#define IT8XXX2_ADC_VCH4DATL     ECREG(IT8XXX2_ADC_BASE + 0x3A)
+#define IT8XXX2_ADC_VCH5CTL      ECREG(IT8XXX2_ADC_BASE + 0x3B)
+#define IT8XXX2_ADC_VCH5DATM     ECREG(IT8XXX2_ADC_BASE + 0x3C)
+#define IT8XXX2_ADC_VCH5DATL     ECREG(IT8XXX2_ADC_BASE + 0x3D)
+#define IT8XXX2_ADC_VCH6CTL      ECREG(IT8XXX2_ADC_BASE + 0x3E)
+#define IT8XXX2_ADC_VCH6DATM     ECREG(IT8XXX2_ADC_BASE + 0x3F)
+#define IT8XXX2_ADC_VCH6DATL     ECREG(IT8XXX2_ADC_BASE + 0x40)
+#define IT8XXX2_ADC_VCH7CTL      ECREG(IT8XXX2_ADC_BASE + 0x41)
+#define IT8XXX2_ADC_VCH7DATM     ECREG(IT8XXX2_ADC_BASE + 0x42)
+#define IT8XXX2_ADC_VCH7DATL     ECREG(IT8XXX2_ADC_BASE + 0x43)
+#define IT8XXX2_ADC_ADCDVSTS     ECREG(IT8XXX2_ADC_BASE + 0x44)
```

Comment:
This seems confusing to me.

Are the ADC register maps so different that we cannot use a struct for them? The pointer indirection in this driver is pretty terrible.

If so, perhaps most registers are at a common offset, but some are not? Then we could just use pointer indirection just for those registers?

Also, does feedback go to the chip-design people to avoid this mess in the future? Registers should have a regular layout.

Reply:
There are three different base address on ADC control registers:
1. ADC0: at offset 0x4
2. ADC 1~3: start from offset 0x6 and offset increases by 3.
3. ADC 4~7: start from offset 0x38 and offset increases by 3.
So we made array (adc_ctrl_regs) to store these registers.
With this array, we won't need to add extra condition branch while controlling an ADC channel.

ADC0 is able to select a specific voltage input channel, so I change to use ADC0 to measure all inputs.

Yes, we have requested chip-design people to fix register layout in the future EC.

Reply:
OK thank you, I think I understand.

Then why not use a shared register struct and just have the base address of each ADC in your tables? Why do you need to store every register's offset, if they are all using the same register format?

E.g. (with made up names)

struct adc_regs {
   uint32_t cr;
   uint32_t sr;
   uint32_t data;
}

struct adc_regs *adc_base[] = {
   IT8XXX2_ADC_BASE + 4,
   IT8XXX2_ADC_BASE + 6,
...
...
   IT8XXX2_ADC_BASE + 0x38,
};

does that work, or are you actually saying every ADC has a different internal register arrangement?

Reply:
> Is that the only difference? If so, you could perhaps add functions to write the control and data registers, with an if...else in each function. It would be more efficient than what you have and easier to understand.
> 
Yes. And Tim have modified to use ADC0 to measure all voltage input channels (0~7). So we don’t need to use ADC1 ~ ADC7’s register now.

> Have you been able to provide feedback to the EC chip-design team about the problems here? I hope it can be fixed for the next generation.

Yes, we have requested chip-design team to fix the registers layout problem in the future EC chip.

Reply:
Thanks